### PR TITLE
[Data] Postpone `reader.get_read_tasks` until execution

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

This PR is to postpone `reader.get_read_tasks()` (i.e. generate the `List[ReadTask]`) until Dataset is executed. Also introduce a hook to allow post processing input files inside `reader.get_read_tasks()`, so we can have custom logic to do post processing of input files, before returning the `List[ReadTask]`.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/tests/test_consumption.py [hunk 1]
```diff
@@ -1413,12 +1413,6 @@ def test_unsupported_pyarrow_versions_check_disabled(
     except ImportError as e:
         pytest.fail(f"_check_pyarrow_version failed unexpectedly: {e}")
 
-    # Test read_parquet.
```
Comment (c21) [inline python/ray/data/tests/test_consumption.py:1416]: Pyarrow 5 does not support pickling the Parquet reader class. Given we do not support Pyarrow 5, remove the test code here. Already verified for Pyarrow 6+, it's not an issue.

File: python/ray/data/read_api.py [hunk 2]
```diff
@@ -384,6 +387,7 @@ def read_datasource(
     # Compute the number of blocks the read will return. If the number of blocks is
     # expected to be less than the requested parallelism, boost the number of blocks
     # by adding an additional split into `k` pieces to each read task.
+    additional_split_factor = None
```
Comment (raulchen) [inline python/ray/data/read_api.py:390]: One small issue. For the new code path, `get_read_tasks` has actually created the read tasks. But the read tasks are only used for the following calculations and then discarded. 
I'm wondering if we can move the following calculation code to the reader. 
So here we only create the reader, but not the read tasks. Also we don't need to expose this `additional_split_factor` to the operator.

File: python/ray/data/tests/test_consumption.py (hunk 1 above)
Comment (raulchen) [reply python/ray/data/tests/test_consumption.py:1416]: should we just remove the PyArrow 5 CI?

**Review by raulchen — APPROVED** [review]

File: python/ray/data/read_api.py (hunk 2 above)
Comment (c21) [reply python/ray/data/read_api.py:390]: yes, that's the major weird thing here. The other code path `LazyBlockList` has all code paths depending on `List[ReadTask]`, so I don't spend time to refactoring `LazyBlockList`. Shall we just delete `DatasetPipeline` and `LazyBlockList`/`BlockList` during 2.8?

File: python/ray/data/tests/test_consumption.py (hunk 1 above)
Comment (c21) [reply python/ray/data/tests/test_consumption.py:1416]: We already removed. CI only tests 6 and 12. This test manually install 5.

File: python/ray/data/read_api.py (hunk 2 above)
Comment (raulchen) [reply python/ray/data/read_api.py:390]: that's fine. can you leave a todo here?

Comment (c21) [reply python/ray/data/read_api.py:390]: SG, added.

File: python/ray/data/_default_config.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# Default file shuffler class to use.
```
Comment (c21) [inline python/ray/data/_default_config.py:1]: I plan to move all of default config value (https://github.com/ray-project/ray/blob/master/python/ray/data/context.py#L17-L133) to this file. But that involves more code change because other code paths directly use them. I plan to do it as a separate PR.

File: python/ray/data/_internal/util.py [hunk 4]
```diff
@@ -513,3 +513,20 @@ def unify_block_metadata_schema(
         # return the first schema.
         return schemas_to_unify[0]
     return None
+
+
+def get_attribute_from_class_name(class_name: str) -> Any:
```
Comment (c21) [inline python/ray/data/_internal/util.py:518]: Checked online, it looks like it's the recommended way to do it - https://stackoverflow.com/questions/452969/does-python-have-an-equivalent-to-java-class-forname .

File: python/ray/data/read_api.py [hunk 5]
```diff
@@ -370,6 +372,7 @@ def read_datasource(
             min_safe_parallelism,
             inmemory_size,
             read_tasks,
+            reader,
         ) = ray.get(
             get_read_tasks.remote(
```
Comment (amogkam) [inline python/ray/data/read_api.py:377]: how is `get_read_tasks` being postponed if it's still being called here?

File: python/ray/data/_internal/logical/operators/read_operator.py [hunk 6]
```diff
@@ -10,18 +10,22 @@ class Read(AbstractMap):
...
+            self._estimated_num_blocks = parallelism
         else:
-            suffix = f"->SplitBlocks({int(estimated_num_blocks / len(read_tasks))})"
+            suffix = f"->SplitBlocks({additional_split_factor})"
+            self._estimated_num_blocks = parallelism * additional_split_factor
         super().__init__(f"Read{datasource.get_name()}{suffix}", None, ray_remote_args)
         self._datasource = datasource
-        self._estimated_num_blocks = estimated_num_blocks
-        self._read_tasks = read_tasks
+        self._reader = reader
```
Comment (amogkam) [inline python/ray/data/_internal/logical/operators/read_operator.py:26]: where does `self._reader` get used here?

File: python/ray/data/datasource/_default_file_post_processor.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+from typing import Any, Dict, List, Tuple
+
+def post_process_files(
+    paths: List[str],
+    file_sizes: List[int],
+    reader_args: Dict[str, Any]
+) -> Tuple[List[str], List[int]]:
```
[the remark this answers is not in the document]
Comment (c21) [reply python/ray/data/datasource/_default_file_post_processor.py]: Sorry but this file is deleted.

File: python/ray/data/read_api.py (hunk 5 above)
Comment (c21) [reply python/ray/data/read_api.py:377]: Added a comment below, this is needed for `LazyBlockList` code path.

File: python/ray/data/_internal/planner/plan_read_op.py [hunk 8]
```diff
@@ -42,7 +42,10 @@ def _plan_read_op(op: Read) -> PhysicalOperator:
     """
 
     def get_input_data() -> List[RefBundle]:
-        read_tasks = op._read_tasks
+        read_tasks = op._reader.get_read_tasks(op._parallelism)
```
Comment (c21) [inline python/ray/data/_internal/planner/plan_read_op.py:45]: >where does self._reader get used here?

This is used in planner here.

File: python/ray/data/_internal/logical/operators/read_operator.py [hunk 9]
```diff
@@ -30,4 +34,4 @@ def fusable(self) -> bool:
         as fusion would prevent the blocks from being dispatched to multiple processes
         for parallel processing in downstream operators.
         """
-        return self._estimated_num_blocks == len(self._read_tasks)
+        return self._parallelism == self._estimated_num_blocks
```
Comment (raulchen) [inline python/ray/data/_internal/logical/operators/read_operator.py:37]: just `self._additional_split_factor is None`?

File: python/ray/data/context.py [hunk 10]
```diff
@@ -204,6 +206,7 @@ def __init__(
         self.use_ray_tqdm = use_ray_tqdm
         self.use_legacy_iter_batches = use_legacy_iter_batches
         self.enable_progress_bars = enable_progress_bars
+        self.file_shuffler = file_shuffler
```
Comment (raulchen) [inline python/ray/data/context.py:209]: nit, there is no need to add add this `__init__` argument. 

```
@property
def file_shuffler_cls(self):
     # import here to avoid cyclic dependencies.
     return DEFAULT_FILE_SHUFFLER;
   
```

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/logical/operators/read_operator.py (hunk 9 above)
Comment (c21) [reply python/ray/data/_internal/logical/operators/read_operator.py:37]: if `_additional_split_factor == 1`, we can still do the fusion right?

File: python/ray/data/_default_config.py [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+DEFAULT_FILE_SHUFFLER = "ray.data.datasource.file_shuffler.SequentialFileShuffler"
```
[the remark this answers is not in the document]
Comment (c21) [reply python/ray/data/_default_config.py:2]: I prefer to store the class name here, because in the runtime, different objects will have different parameters. For example, Spark is doing this - https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L2317-L2322 .

File: python/ray/data/context.py (hunk 10 above)
Comment (c21) [reply python/ray/data/context.py:209]: I guess so, we can change the whole __init__ later. Let's do a separate PR? Just follow other configs here.

File: python/ray/data/datasource/file_shuffler.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+from typing import Any, Dict, List, Tuple
+
+
+class FileShuffler:
```
[the remark this answers is not in the document]
Comment (c21) [reply python/ray/data/datasource/file_shuffler.py]: Sure, updated.

File: python/ray/data/datasource/mongo_datasource.py [hunk 13]
```diff
@@ -104,9 +97,22 @@ def _get_match_query(self, pipeline: List[Dict]) -> Dict:
...
+                self._client, self._database, self._collection
+            )
+            self._avg_obj_size = self._client[self._database].command(
+                "collstats", self._collection
+            )["avgObjSize"]
+
     def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
         from bson.objectid import ObjectId
 
+        self._create_client()
```
[the remark this answers is not in the document]
Comment (c21) [reply python/ray/data/datasource/mongo_datasource.py]: updated.

File: python/ray/data/datasource/file_metadata_shuffler.py [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+class FileMetadataShuffler:
+    """Abstract class for file metadata shuffler.
+
+    Shufflers live on the driver side of the Dataset only.
+    """
+
+    def __init__(self, reader_args: Dict[str, Any]):
+        self._reader_args = reader_args
+
+    def shuffle_files(
```
Comment (scottjlee) [inline python/ray/data/datasource/file_metadata_shuffler.py:13]: nit; similar to class name update, should we update the method name + docstrings to something like `shuffle_file_metadatas` or `shuffle_metadatas`? Is it possibly confusing with shuffling files instead of their metadata?

**Review by scottjlee — APPROVED** [review]
